src/common_subset.rs

```diff
@@ -8,28 +8,33 @@ use std::fmt::{Debug, Display};
 use std::hash::Hash;
 
 use agreement;
-use agreement::Agreement;
+use agreement::{Agreement, AgreementMessage};
 
 use broadcast;
 use broadcast::{Broadcast, TargetedBroadcastMessage};
 
-use proto::{AgreementMessage, BroadcastMessage};
+// FIXME: move into the broadcast module.
+use proto::BroadcastMessage;
 
 // TODO: Make this a generic argument of `Broadcast`.
 type ProposedValue = Vec<u8>;
+// Type of output from the Common Subset message handler.
+type CommonSubsetOutput<NodeUid> = (Option<HashSet<ProposedValue>>, VecDeque<Output<NodeUid>>);
 
 /// Input from a remote node to Common Subset.
 pub enum Input<NodeUid> {
     /// Message from a remote node `uid` to the broadcast instance `uid`.
     Broadcast(NodeUid, BroadcastMessage<ProposedValue>),
-    /// Message from a remote node `uid` to the agreement instance `uid`.
-    Agreement(NodeUid, AgreementMessage),
+    /// Message from a remote node `message_sender_id` concerning the common
+    /// subset element proposed by the node `element_proposer_id`.
+    Agreement {
+        message_sender_id: NodeUid,
```

Comment:
The message sender's ID doesn't need to be part of the message itself, I think, since it's passed into `on_input` anyway.

Reply:
I think you are confusing CommonSubset::on_input which does not receive the ID as argument with Agreement::on_input which does. The latter gets the sender ID from the message received by CommonSubset::on_input.

Reply:
Why can't `CommonSubset::on_input` just pass on the sender ID to `Agreement::on_input`? It looks dangerous to me to have it in the message, since it would basically allow the sender to lie about their identity.

Reply:
Ah, no, the sender ID is in `common_subset::Input` which is not a message but rather a type of input to the Common Subset algorithm. As we discussed before, `AgreementMessage` does not have any IDs.

Reply:
Right, I see! But then what will actually be transmitted over the network, i.e. what's the "common subset" message type? Because that will need to contain `element_proposer_id`, but not `message_sender_id`, I think. So wouldn't it be more natural to replace `Input` with `Message`, and change the signature of `CommonSubset::on_input` so that it takes the sender ID separately from the message?